### ISSUE TITLE
Guidelines for accepting function object parameters in functions modeling loops

### ISSUE THREAD

**Issue by john-lynch** [isocpp/CppCoreGuidelines#128]
I believe we need a guideline for accepting function object parameters in functions that model some form of control flow, and particularly in functions modeling loops of some kind.

Let's start with the example currently in F.24:
```c++
template <class F, class... Args>
inline auto invoke(F&& f, Args&&... args) {
    return forward<F>(f)(forward<Args>(args)...);
}
```
Now let's modify it slightly to bring the issue out:
```c++
template <class F, class... Args>
inline void invoke_n(int n, F&& f, Args&&... args) {
    for (int i = 0; i < n; ++i) {
        forward<F>(f)(forward<Args>(args)...);
    }
}
```
I come across situations like this all the time where I am trying to model some form of loop as a higher order function.

The question is: how should functions like `invoke_n` accept the function object parameter (`f` in this case)?

We can rule out a few options right away. `invoke_n` can't accept `f` as any form of l-value reference or as any form of const reference. A non-const l-value reference would prevent temporary  objects from being passed as the function argument, which is particularly bad for lambdas. Any const reference would prevent mutable function objects from being passed, which is also undesirable.

That leaves us with two options: to accept `f` by forwarding reference, as in the example above, or to accept `f` by value. Both of these have possible issues:

### By forwarding reference ###
```c++
template <class F, class... Args>
inline void invoke_n(int n, F&& f, Args&&... args) {
    for (int i = 0; i < n; ++i) {
        forward<F>(f)(forward<Args>(args)...);
    }
}
```
This has a number of advantages:
1. It accepts any form of function object exactly as we would expect it to.
2. It feels right conceptually: all we're trying to do is just "inject" some code into a different context. This does that as purely as possible.
3. It always forwards its forwarding reference as recommended in the guidelines, so it's a consistent rule to follow.

It also has a possible very odd disadvantage: it forwards its function object (and indeed, its parameter pack) multiple times. This feels really dirty, since it means that we will be (possibly) using a single object as an r-value over and over again. This is potentially problematic since we can ref qualify functions and the writer of a function object class may not expect an r-value qualified member function to be called repeatedly. This is likely not an issue for lambdas, but could be for other function objects. Indeed, one could see extending this issue to simply ask about forwarding *any* kind of object multiple times (one could envision a function like `make_n` that constructs `n` copies of some type and returns them in a `vector`, for example; indeed, this very example forwards its parameter pack multiple times, which could get hairy).

There's another option, of course:
```c++
template <class F, class... Args>
inline void invoke_n(int n, F&& f, Args&&... args) {
    for (int i = 0; i < n; ++i) {
        f(forward<Args>(args)...);
    }
}
```
This avoids the problem of forwarding `f` multiple times (though it still forwards its parameter pack multiple times). On the other hand, it treats a possible r-value reference like an l-value and it creates a corner case for an otherwise consistent rule: here we are *not* forwarding a forwarding reference. This also feels gross.

### By value ###
```c++
template <class F, class... Args>
inline void invoke_n(int n, F f, Args&&... args) {
    for (int i = 0; i < n; ++i) {
        f(forward<Args>(args)...);
    }
}
```
This avoids all the reference issues above, but has two major downsides:
1. It relies heavily on the optimizer's inlining ability to elide copies. While perhaps a reasonable expectation, it doesn't feel great to copy a function object just to model a loop.
2. It won't work with any function object that wants to mutate itself and allow the caller to read those results. That's not an issue with lambdas, but it would be with a good deal of plausible hand-rolled function objects.

In short, I'm stumped as to how to handle this case. I guess my knee jerk reaction is to simply forward every forwarding reference every time, but I can see that causing problems. Some guidance on this issue would be greatly appreciated, and it may even need to be extended to guidance for more than just function objects; forwarding the parameter pack in this example is equally problematic. It's the function object that I find most interesting, however.

**Comment by AndrewPardoe** [comment]
We've discussed this and we think the right answer is to pass f by value and to not forward it. You're not forwarding it along to something else like you do the arguments, and the guidance already for passing function objects is to treat them like function pointers--pass them by value.

**Comment by john-lynch** [comment]
I have no problem with this advice, but it does raise a couple of other issues:
1. I may simply be a poor reader, but I can't find any guidance that say to pass function objects like function pointers. If this guidance is intended, it should be made explicit, preferably in section "F.call: Argument passing."
2. Why does the example in F.24 accept its function object by forwarding reference if function objects should be accepted like function pointers, that is, by value?
3. The `invoke_n` example I gave above still forwards its parameter pack multiple times. Is that desirable (or even legal)? What advice should we give on that front? I can easily imagine a scenario along these lines:
```c++
void bar(vector<int>&& v);
vector<int> calculate_some_expensive_vector();
void foo(int n)
{
    invoke_n(n, bar, calculate_some_expensive_vector());
}
```
This code has issues because of the fact that we forward the parameter pack inside `invoke_n` multiple times. It will compile, but the first invocation of `bar` will likely move out of the expensive vector, rendering subsequent invocations problematic.

I think a large part of the issue here is that the behavior of invoke_n doesn't match our intuition from any way we might express this loop in "normal" C++ code:
```c++
void bar(vector<int>&& v);
vector<int> calculate_some_expensive_vector();
void foo(int n)
{
    for (int i = 0; i < n; ++i) {
        bar(calculate_some_expensive_vector());
    }
}
```
This version is expensive, since it calculates the vector multiple times, but it will (likely) behave as desired.
```c++
void bar(vector<int>&& v);
vector<int> calculate_some_expensive_vector();
void foo(int n)
{
    auto v = calculate_some_expensive_vector();
    for (int i = 0; i < n; ++i) {
        bar(v);
    }
}
```
This version calculates the vector only once, which is nice. Of course, it won't compile, because you can't convert an l-value to an r-value reference. We'd have to explicitly move from it, which would be an obvious bug, or do something like the following:
```c++
void bar(vector<int>&& v);
vector<int> calculate_some_expensive_vector();
void foo(int n)
{
    auto v = calculate_some_expensive_vector();
    for (int i = 0; i < n; ++i) {
        auto v_copy = v;
        bar(std::move(v_copy));
    }
}
```
Whether this is preferable to calling `calculate_some_expensive_vector` multiple times inside the loop body is up to the writer of `foo`. 

I think functions that try to abstract control flow (like `invoke_n`) shouldn't violate our intuitions like this. Forwarding the parameter pack (or any other forwarding reference) multiple times basically short-circuits the safety mechanisms in place in the type system to prevent these violations. Is there anyway to write a function like `invoke_n` so that it matches our intuition? If not, perhaps we should add guidance indicating that parameters should never be forwarded multiple times? This comes with limitations of its own, but will not silently accept buggy code.

**Comment by AndrewPardoe** [comment]
Discussed 11/30/2015.

**Comment by hsutter** [comment]
I've expanded the Enforcement in F.19 to clarify the "forward exactly once, and don't do anything else with it" guidance.

**Comment by hsutter** [comment]
And I've added a note in SS-lambdas that function objects should be cheap to copy and so passed by value (with xref to Rf-in). Thanks!
